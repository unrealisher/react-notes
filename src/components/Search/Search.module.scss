.search {
  min-height: 32px;
  display: flex;
  background-color: #e6e6e6;
  border-radius: 4px;
  align-items: center;
}

.wrapper {
  margin-left: 2px;
  flex-grow: 1;
  flex-shrink: 1;
  position: relative;
}

.input {
  min-height: 29px;
  padding-left: 20px;
  padding-right: 30px;
  width: 100%;
  box-sizing: border-box;
  border: none;
  border-radius: 2px;
  font-size: 15px;
  line-height: 18px;
}

.reset {
  cursor: pointer;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 6px);
  right: 8px;

  &::before,
  &::after {
    content: "";
    position: absolute;
    width: 8px;
    height: 1px;
    background-color: white;
    top: calc(50% - 0.5px);
    left: 2px;
  }

  &::before {
    transform: rotate(45deg);
  }

  &::after {
    transform: rotate(-45deg);
  }

  &:hover,
  &:focus {
    background-color: rgba(51, 51, 51, 1);
  }

  &:active {
    background-color: rgba(51, 51, 51, 0.8);
  }
}

.button {
  cursor: pointer;
  background-color: inherit;
  padding: 6px 10px;
  font-size: 15px;
  line-height: 18px;
}

@media (min-width: 970px) {
  .search {
    position: absolute;
    left: calc(50% - 280px);
    top: 15px;
    width: 440px;
  }

  .input {
    box-sizing: border-box;
    width: 364px;
  }

  .button {
    width: 76px;
    box-sizing: border-box;
  }
}

@media (min-width: 1120px) {
  .search {
    left: calc(50% - 238px);
  }
}
