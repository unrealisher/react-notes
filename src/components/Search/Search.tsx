import React, { useRef, useState } from "react";

import styles from "./Search.module.scss";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { setSearchItems } from "../../store/actions/searchItems";
import IState from "../../interfaces/IState";

interface IStateToProps {
  search: string;
}

interface IDispatchToProps {
  onTextChange: Function;
}

interface IProps extends IStateToProps, IDispatchToProps {}

export const Search = (props: IProps): JSX.Element => {
  const { search, onTextChange } = props;
  const refInput = useRef<HTMLInputElement>(null);
  const [value, setValue] = useState(search);
  return (
    <article className={styles.search}>
      <div className={styles.wrapper}>
        <input
          ref={refInput}
          className={styles.input}
          type="text"
          placeholder="Поиск"
          onChange={() => {
            if (refInput.current) setValue(refInput.current.value);
          }}
          onKeyUp={event => {
            if (event.keyCode === 13 && refInput.current)
              onTextChange(refInput.current.value);
          }}
          value={value}
        />
        <button
          className={styles.reset}
          onClick={() => {
            setValue("");
            onTextChange("");
          }}
        />
      </div>
      <button
        className={styles.button}
        onClick={() => {
          if (refInput.current) onTextChange(refInput.current.value);
        }}
      >
        Поиск
      </button>
    </article>
  );
};

const mapStateToProps = (state: IState): IStateToProps => {
  return {
    search: state.search
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchToProps => {
  return {
    onTextChange: (text: string): void => {
      dispatch(setSearchItems(text));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Search);
