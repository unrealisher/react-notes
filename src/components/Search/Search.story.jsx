import React from "react";
import { Provider } from "react-redux";
import store from "./../../store/index";

import Search from "./Search";

import { storiesOf } from "@storybook/react";

import "./../../index.scss";

storiesOf("Search", module).add("Стандарт", () => (
  <Provider store={store}>
    <Search />
  </Provider>
));
