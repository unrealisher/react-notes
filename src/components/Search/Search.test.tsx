import React from "react";
import { shallow } from "enzyme";
import { Search } from "./Search";

describe("Search", () => {
  const search = "";
  const onTextChange = (): void => {};
  it("correct render", () => {
    const component = shallow(
      <Search search={search} onTextChange={onTextChange} />
    );
    expect(component).toMatchSnapshot();
  });
});
