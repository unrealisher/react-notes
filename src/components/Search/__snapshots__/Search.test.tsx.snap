// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search correct render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    onTextChange={[Function]}
    search=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="wrapper"
        >
          <input
            className="input"
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Поиск"
            type="text"
            value=""
          />
          <button
            className="reset"
            onClick={[Function]}
          />
        </div>,
        <button
          className="button"
          onClick={[Function]}
        >
          Поиск
        </button>,
      ],
      "className": "search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="input"
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Поиск"
              type="text"
              value=""
            />,
            <button
              className="reset"
              onClick={[Function]}
            />,
          ],
          "className": "wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input",
              "onChange": [Function],
              "onKeyUp": [Function],
              "placeholder": "Поиск",
              "type": "text",
              "value": "",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "reset",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Поиск",
          "className": "button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Поиск",
        "type": "button",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="wrapper"
          >
            <input
              className="input"
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Поиск"
              type="text"
              value=""
            />
            <button
              className="reset"
              onClick={[Function]}
            />
          </div>,
          <button
            className="button"
            onClick={[Function]}
          >
            Поиск
          </button>,
        ],
        "className": "search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="input"
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="Поиск"
                type="text"
                value=""
              />,
              <button
                className="reset"
                onClick={[Function]}
              />,
            ],
            "className": "wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "input",
                "onChange": [Function],
                "onKeyUp": [Function],
                "placeholder": "Поиск",
                "type": "text",
                "value": "",
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "reset",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Поиск",
            "className": "button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Поиск",
          "type": "button",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
